File: desktop/src/shared/burst.ts
// Adag-szabály oldalanként: ennyi HASZNÁLAT után ennyi SZÜNET.
//
// A napi keret testvére, más alakú lyukra: a keret a napi összesenről szól,
// ez arról, hogy egyszerre mennyi fér. A példa, amiből született:
// két perc Gemini után tíz perc tiltás, aztán az oldal magától feloldódik.
// Nem büntetés, hanem ütem: a rövid odapillantás belefér, a belefeledkezés
// nem.
//
// HOGYAN SZÁMOL. A mérés kötegekben érkező mintáiból oldalanként adag-számláló
// gyűlik. Ha a számláló eléri az adagot, indul a hűtés (DNS-tiltás), és a
// számláló nullázódik. Ha egy hűtésnyi ideig nem használtad az oldalt, a
// számláló magától tiszta lappal indul — e nélkül a hetekkel korábbi fél
// percek is összeadódnának, és a tiltás az égből esne az emberre.
//
// A SZÁMLÁLÓ ESZKÖZ-HELYI. A beállítás (adag + szünet hossza) a blokklista
// rekordján utazik a többi gépre, a számláló és a hűtés vége viszont nem:
// a szinkron tízperces körökben jár, egy kétperces adaghoz az túl lassú —
// ebből nem pontatlan közös számláló lesz, hanem őszintén eszközönkénti.
//
// Tiszta és függőség nélküli, mint a megosztott mag többi része — a Kotlin
// tükrözi (Android is mér és tilt), a Swift a mezőket viszi át (az iPhone nem
// mér előteret, ott a szabály nem érvényesül — kimondva, nem elhallgatva).
// Lásd docs/feature-burst-limit.md.

/** Az adag legfeljebb egy nap — ami ennél több, az már a napi keret dolga. */
export const MAX_BURST_MINUTES = 24 * 60;
/** A szünet is legfeljebb egy nap — ami hosszabb, az sima tiltás legyen. */
export const MAX_COOLDOWN_MINUTES = 24 * 60;

export interface BurstRule {
  /** ennyi aktív használat fér egy adagba, másodpercben */
  burstSeconds: number;
  /** ennyi ideig tilt utána, másodpercben */
  cooldownSeconds: number;
}

/**
 * Az adag-számláló egy oldalra, EZEN a gépen.
 *
 * Nem kerül a drótra: lásd a fejlécet. A `lastAt` a legutóbb elszámolt minta
 * ideje — ebből derül ki, hogy volt-e egy hűtésnyi pihenő.
 */
export interface BurstState {
  usedSeconds: number;
  lastAt: number;
  cooldownUntil: number;
}

/**
 * Használható szabály a két beírt értékből, vagy null.
 *
 * A kettő CSAK EGYÜTT értelmes: adag szünet nélkül nem tilt semmit, szünet
 * adag nélkül sosem indul el. A fél-kitöltött állapotot ezért nem tároljuk el
 * félig, hanem nincs szabály.
 */
export function normalizeBurst(
  burstSeconds: number | undefined | null, cooldownSeconds: number | undefined | null,
): BurstRule | null {
  if (burstSeconds === undefined || burstSeconds === null) return null;
  if (cooldownSeconds === undefined || cooldownSeconds === null) return null;
  if (!Number.isFinite(burstSeconds) || burstSeconds <= 0) return null;
  if (!Number.isFinite(cooldownSeconds) || cooldownSeconds <= 0) return null;
  return {
    burstSeconds: Math.min(Math.round(burstSeconds), MAX_BURST_MINUTES * 60),
    cooldownSeconds: Math.min(Math.round(cooldownSeconds), MAX_COOLDOWN_MINUTES * 60),
  };
}

/**
 * Egy elfogadott mérés-minta elszámolása az adag-számlálóban.
 *
 * A minták időbélyeggel jönnek, és nem feltétlenül sorban — a köteg múltbeli
 * szeleteket is hozhat. Ezért:
 *
 *   - hűtés alatt a minta NEM számít (a tiltott oldal hibalapján ülve mért
 *     másodpercek különben újraindítanák a hűtést a lejárta után);
 *   - ha a minta és az előző elszámolt minta közt egy hűtésnyi csend volt,
 *     a számláló tiszta lappal indul;
 *   - a `lastAt` sosem lép hátra: egy elkésett régi minta nem gyárthat
 *     hamis pihenőt.
 *
 * A visszaadott állapot ÚJ objektum — a hívó dönt, hova teszi.
 */
export function noteBurstUsage(
  rule: BurstRule, st: BurstState | undefined | null, seconds: number, at: number,
): BurstState {
  const cur: BurstState = st ?? { usedSeconds: 0, lastAt: 0, cooldownUntil: 0 };
  if (!Number.isFinite(seconds) || seconds <= 0 || !Number.isFinite(at)) return cur;
  if (cur.cooldownUntil > at) return cur;
  let used = cur.usedSeconds;
  if (cur.lastAt > 0 && at - cur.lastAt >= rule.cooldownSeconds * 1000) used = 0;
  used += seconds;
  const lastAt = Math.max(cur.lastAt, at);
  if (used >= rule.burstSeconds) {
    // Az adag betelt: indul a hűtés, a számláló nulláról jön vissza. A hűtés
    // a MINTA idejétől számít, nem a beérkezésétől — ha a köteg késett, a
    // különbség már le is telt belőle.
    return { usedSeconds: 0, lastAt, cooldownUntil: at + rule.cooldownSeconds * 1000 };
  }
  return { usedSeconds: used, lastAt, cooldownUntil: cur.cooldownUntil };
}

/** Tart-e most a hűtés. */
export function isCoolingDown(st: BurstState | undefined | null, now: number): boolean {
  return !!st && st.cooldownUntil > now;
}

/**
 * LAZÍTÁS-e a szabály cseréje — mert a lazítás próbatételbe kerül.
 *
 * Szigorítás (ingyen): szabály felvétele, kisebb adag, hosszabb szünet.
 * Lazítás (próbatétel): a szabály levétele, nagyobb adag, rövidebb szünet.
 * Vegyes módosításnál elég egyetlen lazító irány, és már próbatétel — ahogy
 * a menetrendnél is a lazító fele dönt.
 */
export function isBurstLoosening(current: BurstRule | null, next: BurstRule | null): boolean {
  if (current === null) return false;      // nem volt szabály: bármilyen szabály szigorúbb
  if (next === null) return true;          // a szabály levétele felszabadítja az oldalt
  return next.burstSeconds > current.burstSeconds
    || next.cooldownSeconds < current.cooldownSeconds;
}

File: android/app/src/main/java/hu/breaker/app/core/Burst.kt
package hu.breaker.app.core

/**
 * Adag-szabály oldalanként — a desktop/src/shared/burst.ts tükre.
 *
 * Ennyi HASZNÁLAT után ennyi SZÜNET: például 2 perc Gemini után 10 perc
 * tiltás, aztán az oldal magától kinyílik. A napi keret testvére, más alakú
 * lyukra: a keret a napi összesenről szól, ez arról, hogy egyszerre mennyi
 * fér.
 *
 * A BEÁLLÍTÁS (adag + szünet) a blokklista rekordján utazik a többi eszközre;
 * a SZÁMLÁLÓ eszköz-helyi. A szinkron tízperces körökben jár, egy kétperces
 * adaghoz az túl lassú — ebből nem pontatlan közös számláló lesz, hanem
 * őszintén eszközönkénti. Ha itt változtatsz, a TS ikren is.
 */
object BurstLogic {

    /** Az adag legfeljebb egy nap — ami több, az már a napi keret dolga. */
    const val MAX_BURST_MINUTES = 24 * 60
    /** A szünet is legfeljebb egy nap — ami hosszabb, az sima tiltás legyen. */
    const val MAX_COOLDOWN_MINUTES = 24 * 60

    data class Rule(val burstSeconds: Long, val cooldownSeconds: Long)

    /**
     * Az adag-számláló egy oldalra, EZEN a készüléken. Nem kerül a drótra.
     * A [lastAt] a legutóbb elszámolt minta ideje — ebből derül ki, hogy
     * volt-e egy hűtésnyi pihenő.
     */
    data class State(
        val usedSeconds: Double = 0.0,
        val lastAt: Long = 0,
        val cooldownUntil: Long = 0,
    )

    /**
     * Használható szabály a két értékből, vagy null. A kettő CSAK EGYÜTT
     * értelmes: adag szünet nélkül nem tilt semmit, szünet adag nélkül sosem
     * indul el — fél-kitöltött állapotot nem tárolunk.
     */
    fun normalize(burstSeconds: Long?, cooldownSeconds: Long?): Rule? {
        if (burstSeconds == null || cooldownSeconds == null) return null
        if (burstSeconds <= 0 || cooldownSeconds <= 0) return null
        return Rule(
            burstSeconds.coerceAtMost(MAX_BURST_MINUTES * 60L),
            cooldownSeconds.coerceAtMost(MAX_COOLDOWN_MINUTES * 60L),
        )
    }

    /**
     * Egy mérés-minta elszámolása az adag-számlálóban.
     *
     *   - hűtés alatt a minta NEM számít (a tiltott oldalon mért idő különben
     *     újraindítaná a hűtést a lejárta után);
     *   - ha a minta és az előző közt egy hűtésnyi csend volt, a számláló
     *     tiszta lappal indul;
     *   - a lastAt sosem lép hátra: egy elkésett régi minta nem gyárthat
     *     hamis pihenőt.
     */
    fun noteUsage(rule: Rule, st: State?, seconds: Double, at: Long): State {
        val cur = st ?: State()
        if (!seconds.isFinite() || seconds <= 0) return cur
        if (cur.cooldownUntil > at) return cur
        var used = cur.usedSeconds
        if (cur.lastAt > 0 && at - cur.lastAt >= rule.cooldownSeconds * 1000) used = 0.0
        used += seconds
        val lastAt = maxOf(cur.lastAt, at)
        if (used >= rule.burstSeconds) {
            // Az adag betelt: indul a hűtés, a számláló nulláról jön vissza.
            // A hűtés a MINTA idejétől számít — ha a köteg késett, a
            // különbség már le is telt belőle.
            return State(0.0, lastAt, at + rule.cooldownSeconds * 1000)
        }
        return State(used, lastAt, cur.cooldownUntil)
    }

    /** Tart-e most a hűtés. */
    fun isCoolingDown(st: State?, now: Long): Boolean =
        st != null && st.cooldownUntil > now

    /**
     * Lazítás-e a szabály cseréje — mert a lazítás próbatételbe kerül.
     * Szigorítás (ingyen): felvétel, kisebb adag, hosszabb szünet. Lazítás:
     * levétel, nagyobb adag, rövidebb szünet — vegyesnél a lazító fele dönt.
     */
    fun isLoosening(current: Rule?, next: Rule?): Boolean {
        if (current == null) return false
        if (next == null) return true
        return next.burstSeconds > current.burstSeconds ||
            next.cooldownSeconds < current.cooldownSeconds
    }
}
